### PR TITLE
#540 BigDecimal equality using compareTo

### PR DESCRIPTION
**Pull request by ac183**
# What problem does this pull request solve?
Issue #540 

# Please provide any additional information below.

# NOTE
* Please run `mvn spotless:apply` to format the code before opening a PR. Otherwise, GitHub Actions will complain at you 😉. Unfortunately, you will need to have Node installed to do so.
* Mutation tests will be run by [PITest](https://pitest.org/) after opening the PR. It will post comments in the PR for each issue found. Please take a look at them, but if the comments don't make sense, please don't worry about them.

### CODE REVIEW

**Comment by jqno** [comment]
Looks really nice!

I think, now the new checks are default behavior, that the manual entry makes more sense on the error messages page, since that is probably where people will start looking when they run into this behaviour.

Also I'm not sure what's going on with the mutation tests, but I haven't had time to investigate yet.

**Comment by ac183** [comment]
2 of the mutants are for the tests not catching if the equals checking in `BigDecimalFieldCheck` was removed.

The other one where +1 can be replaced with -1 is true but +1, -1, +2, ... should all be valid implementations to change the scale of the `BigDecimal` field to make it different for `equals` and `hashCode` but still the same for `compareTo`.

I'll follow up with relocating the doc.

Let me know if you prefer the commits to be squashed, whether I should wait until all review and changes done before squashing, anything about commit messages, etc.

**Comment by jqno** [comment]
I have no strict rules about committing, so please do what feels best to you. Thanks for following up. If there's no good way to deal with a mutant, then that's ok. The check is there to help avoid bugs, no need to be super strict about it.

**Comment by jqno** [comment]
Hi! I did some work on the website this weekend. I don't think it should cause merge conflicts for you, but if it does, please just push your changes when you're ready and let me know, then I will fix any conflicts that might occur.

**Comment by ac183** [comment]
No conflicts :)

Updating the doc threw up a bit of a puzzle. Given this warning is raised, it would be nice to advise how it should be solved. I already have the answer of using `compareTo` and having to "normalise" the hashcode, hence the feature request and PR. The hashcode normalisation is open to different implementations. Calling `stripTrailingZeros` before hashing seems fine. I've seen it in a few posts online and looked at the code to confirm: yes it does work generally e.g. for different `BigDecimal` scales.

As it has a cost, another option is performing this call up front in the constructor. I don't think a check can be created to see if someone has done this without them providing extra config because you need to actually call the constructor. This was annoying and led to some thinking about is it ok to add a warning when 1 of 2 solutions cannot be checked. My conclusion is as conveyed in the `bigdecimal-equality.md` error message doc. So I think firstly it is useful to warn people to consider what to do if they're using `BigDecimal` and secondly, really, `BigDecimal` is not fit for purpose in terms of equality - without being able to go so far as knowing or implementing whatever the best substitute is. And there are 2 ways (at least) to solve if you *are* using it (and care about variants being equal). I *speculate* that for most people the `compareTo` choice over normalisation in the constructor is fine and makes sense out of the 2 if you're not hashing lots (mentioned in doc).

File: docs/_pages/errormessages.md [hunk 1]
```diff
@@ -19,13 +19,14 @@ Before you continue, please check your class's `toString` method. Since EqualsVe
...
 * [Abstract delegation](/equalsverifier/errormessages/abstract-delegation)
+* [BigDecimal equality](/equalsverifier/errormessages/bigdecimal-equality)
 * [ClassCastException: java.lang.Object cannot be cast to …](/equalsverifier/errormessages/classcastexception)
 * [Coverage is not 100%](/equalsverifier/errormessages/coverage-is-not-100-percent)
 * [Double: equals doesn't use Double.compare for field foo](/equalsverifier/errormessages/double-equals-doesnt-use-doublecompare-for-field-foo)
 * [Float: equals doesn't use Float.compare for field foo](/equalsverifier/errormessages/float-equals-doesnt-use-floatcompare-for-field-foo)
 * [Mutability: equals depends on mutable field](/equalsverifier/errormessages/mutability-equals-depends-on-mutable-field)
 * [NoClassDefFoundError](/equalsverifier/errormessages/noclassdeffounderror)
-* [Non-nullity: equals/hashCode/toString throws NullPointerExcpetion](/equalsverifier/errormessages/non-nullity-equals-hashcode-tostring-throws-nullpointerexception)
+* [Non-nullity: equals/hashCode/toString throws NullPointerException](/equalsverifier/errormessages/non-nullity-equals-hashcode-tostring-throws-nullpointerexception)
```
Comment (jqno) [inline docs/_pages/errormessages.md:29]: Ha, you also fixed a typo! Thanks! :)

**Review by jqno — APPROVED** [review]

**Comment by jqno** [comment]
Thanks for the detailed documentation. I understand your struggle, but I agree with your conclusions. There is only so much EqualsVerifier can reasonably do, and if that's not sufficient, you can always suppress the warning. So thank you again for the well-thought out PR. I'll merge it now, but there's another thing I'd to fix before releasing -- hopefully I can get that done today or Monday. I'll let you know when I do the release!

**Comment by ac183** [comment]
Thanks for the prompt, friendly and helpful engagement. Positive experience for my first open source contribution :)

**Comment by jqno** [comment]
It was a positive experience for me as well!

**Pull request merged**